tensorflow/core/kernels/roll_op_gpu.cu.cc

```diff
@@ -0,0 +1,103 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/kernels/roll_op.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+
+template <typename T>
+__global__ void RollKernel(
+    const int32 nthreads, const int32 num_dims,
+    const T* input, T* output,
+    const int32* dim_size, const int32* threshold, const int64* dim_range) {
+  CUDA_1D_KERNEL_LOOP(out_idx, nthreads) {
+    int64 offset = 0;
+    for (int i = 0; i < num_dims; i++) {
+        const int64 stride = dim_range[i] / dim_size[i];
```

Comment:
That's a lot of index computation to move a single element. I'm pretty sure we can do better, but let's keep it simple for now.